File: ios/SsdidWallet/Domain/Did/DidJwkResolver.swift
import Foundation

final class DidJwkResolver: DidResolver {
    func resolve(did: String) async throws -> DidDocument {
        guard did.hasPrefix("did:jwk:") else {
            throw DidResolutionError.unsupportedMethod(did)
        }
        let encoded = String(did.dropFirst("did:jwk:".count))
        guard Data(base64Encoded: base64UrlToBase64(encoded)) != nil else {
            throw DidResolutionError.invalidJwk
        }
        let keyId = "\(did)#0"
        let vm = VerificationMethod(
            id: keyId, type: "JsonWebKey2020", controller: did, publicKeyMultibase: ""
        )
        return DidDocument(
            id: did, controller: did, verificationMethod: [vm],
            authentication: [keyId], assertionMethod: [keyId], capabilityInvocation: [keyId]
        )
    }

    private func base64UrlToBase64(_ input: String) -> String {
        var s = input.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        let remainder = s.count % 4
        if remainder > 0 { s += String(repeating: "=", count: 4 - remainder) }
        return s
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/did/DidJwkResolver.kt
package my.ssdid.wallet.domain.did

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import my.ssdid.wallet.domain.model.DidDocument
import my.ssdid.wallet.domain.model.VerificationMethod
import java.util.Base64

class DidJwkResolver : DidResolver {
    override suspend fun resolve(did: String): Result<DidDocument> = runCatching {
        require(did.startsWith("did:jwk:")) { "Not a did:jwk: $did" }
        val encoded = did.removePrefix("did:jwk:")
        val decodedBytes = Base64.getUrlDecoder().decode(encoded)
        val jwkString = String(decodedBytes, Charsets.UTF_8)
        val jwkObject = Json.parseToJsonElement(jwkString).jsonObject
        val keyId = "$did#0"
        val vm = VerificationMethod(
            id = keyId,
            type = "JsonWebKey2020",
            controller = did,
            publicKeyMultibase = "",
            publicKeyJwk = jwkObject
        )
        DidDocument(
            id = did,
            controller = did,
            verificationMethod = listOf(vm),
            authentication = listOf(keyId),
            assertionMethod = listOf(keyId),
            capabilityInvocation = listOf(keyId)
        )
    }
}
